### PR TITLE
storage: skip scanning lock when using rc

### PR DESCRIPTION
**Pull request by you06**
Signed-off-by: you06 <you1474600@gmail.com>

### What is changed and how it works?

Fix RC for #11485

What's Changed:

If the isolation level is RC, the scanner read lock and do nothing with it, this PR let RC skip reading locks.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- PR to update `pingcap/tidb-ansible`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

Side effects

### Release note 

```release-note
Improve performance for scanning in RC by skipping reading lock.
```

### CODE REVIEW

File: src/storage/mvcc/reader/scanner/forward.rs [hunk 1]
```diff
@@ -592,23 +600,22 @@ fn scan_latest_handle_lock<S: Snapshot, T>(
...
-            };
-            Lock::check_ts_conflict(
-                Cow::Owned(lock),
-                &current_user_key,
-                cfg.ts,
-                &cfg.bypass_locks,
-            )
-        }
-        IsolationLevel::Rc => Ok(()),
+    if cfg.isolation_level == IsolationLevel::Rc {
```
Comment (youjiali1995) [inline src/storage/mvcc/reader/scanner/forward.rs:603]: I will code like this
```rust
    if cfg.isolation_level == IsolationLevel::Rc {
        return Ok(HandleRes::Skip(current_user_key));
    }
    // Only needs to check lock in SI
    let lock_cursor = cursors.lock.as_mut().unwrap();
    let lock = {
        let lock_value = lock_cursor.value(&mut statistics.lock);
        Lock::parse(lock_value)?
    };
    lock_cursor.next(&mut statistics.lock);

    Lock::check_ts_conflict(
        Cow::Owned(lock),
        &current_user_key,
        cfg.ts,
        &cfg.bypass_locks,
    )
    .or_else(|e| {
        // Even if there is a lock error, we still need to step the cursor for future
        // calls.
        statistics.lock.processed_keys += 1;
        cursors
            .move_write_cursor_to_next_user_key(&current_user_key, statistics)
            .and(Err(e.into()))
    })
    .map(|_| HandleRes::Skip(current_user_key))
```

**Review by youjiali1995 — COMMENTED** [review]
rest lgtm

**Comment by youjiali1995** [comment]
BTW, it's better to add an integration test to guarantee the isolation level in `kvrpcpb::Context` is passed to the scanner.

**Review by sticnarf — COMMENTED** [review]
I don't have other comments. The rest lgtm.

File: src/storage/mvcc/reader/scanner/mod.rs [hunk 2]
```diff
@@ -978,4 +985,47 @@ mod tests {
...
+        }
+
+        must_prewrite_put(&engine, key1, val1, key1, 10);
+        must_commit(&engine, key1, 10, 20);
+
+        must_prewrite_put(&engine, key2, val2, key2, 30);
+        must_commit(&engine, key2, 30, 40);
+
+        // use async commit to avoid skipping lock.
+        must_prewrite_put_async_commit(&engine, key1, val12, key1, &Some(vec![]), 50, 0);
```
[the remark this answers is not in the document]
Comment (you06) [reply src/storage/mvcc/reader/scanner/mod.rs:1011]: SI reads with `u64::MAX` will ignore non-async-commit locks. Here using async commit for making this case fail in SI.

File: src/storage/mvcc/reader/scanner/forward.rs [hunk 3]
```diff
@@ -364,15 +372,15 @@ impl<S: Snapshot> ScanPolicy<S> for LatestKvPolicy {
         statistics: &mut Statistics,
     ) -> Result<HandleRes<Self::Output>> {
         if cfg.isolation_level == IsolationLevel::Rc {
-            cursors.lock.next(&mut statistics.lock);
```
[the remark this answers is not in the document]
Comment (you06) [reply src/storage/mvcc/reader/scanner/forward.rs:367]: Yes, `IsolationLevel::Rc` should not enter this function.

**Review by youjiali1995 — APPROVED** [review]

**Comment by youjiali1995** [comment]
@sticnarf PTAL again.

**Review by sticnarf — APPROVED** [review]

**Pull request merged**